source/Calamari.Terraform/TerraformCliExecutor.cs

```diff
@@ -0,0 +1,212 @@
+            Workspace = variables.Get(TerraformSpecialVariables.Action.Terraform.Workspace);
+            InitParams = variables.Get(TerraformSpecialVariables.Action.Terraform.AdditionalInitParams);
+            ActionParams = variables.Get(TerraformSpecialVariables.Action.Terraform.AdditionalActionParams);
+            TerraformExecutable = variables.Get(TerraformSpecialVariables.Action.Terraform.CustomTerraformExecutable) ??
+                                  $"terraform{(CalamariEnvironment.IsRunningOnWindows ? ".exe" : String.Empty)}";
+            AllowPluginDownloads = variables.GetFlag(TerraformSpecialVariables.Action.Terraform.AllowPluginDownloads, true);
+            AttachLogFile = variables.GetFlag(TerraformSpecialVariables.Action.Terraform.AttachLogFile);
+            TerraformVariableFiles = GenerateVarFiles();
+
+            logPath = Path.Combine(deployment.CurrentDirectory, "terraform.log");
+
+            if (!String.IsNullOrEmpty(TemplateDirectory))
+            {
+                var templateDirectoryTemp = Path.Combine(deployment.CurrentDirectory, TemplateDirectory);
+
+                if (!Directory.Exists(templateDirectoryTemp))
+                {
+                    throw new Exception($"Directory {TemplateDirectory} does not exist.");
+                }
+
+                TemplateDirectory = templateDirectoryTemp;
+            }
+
+            InitialiseTerraformEnvironmentVariables();
+
+            InitialisePlugins();
+
+            InitialiseWorkspace();
+        }
+
+        public string ExecuteCommand(params string[] arguments)
+        {
+            var commandResult = ExecuteCommandInternal(ToSpaceSeparated(arguments), out var result);
+
+            commandResult.VerifySuccess();
+
+            return result;
+        }
+
+        public int ExecuteCommand(out string result, params string[] arguments)
+        {
+            var commandResult = ExecuteCommandInternal(ToSpaceSeparated(arguments), out result);
+
+            return commandResult.ExitCode;
+        }
+
+        public int ExecuteCommand(out string result, out CommandResult commandResult, params string[] arguments)
+        {
+            commandResult = ExecuteCommandInternal(ToSpaceSeparated(arguments), out result);
+
+            return commandResult.ExitCode;
+        }
+
+        public void Dispose()
+        {
+            if (AttachLogFile)
+            {
+                if (fileSystem.FileExists(logPath))
+                {
+                    Log.NewOctopusArtifact(fileSystem.GetFullPath(logPath), fileSystem.GetFileName(logPath), fileSystem.GetFileSize(logPath));
+                }
+            }
+        }
+
+        static string ToSpaceSeparated(IEnumerable<string> items)
+        {
+            return string.Join(" ", items.Where(_ => !String.IsNullOrEmpty(_)));
+        }
+
+        CommandResult ExecuteCommandInternal(string arguments, out string result)
+        {
+            var environmentVar = defaultEnvironmentVariables;
```

Comment:
Since we initialize the `defaultEnvironmentVariables` variable in the contructor, do we ever need to check if its null?

Reply:
For public apis I would definitely, but given this is an internal api I don't tend to do it

Reply:
right sorry, mis-read `defaultEnvironmentVariables` as `enviromentVariables`